client/gutenberg/editor/calypsoify-iframe.tsx

```diff
@@ -38,28 +37,57 @@ import { getEditorPostId } from 'state/ui/editor/selectors';
  */
 import './style.scss';
 
-class CalypsoifyIframe extends Component {
-	static propTypes = {
-		postId: PropTypes.number,
-		postType: PropTypes.string,
-		duplicatePostId: PropTypes.number,
-		pressThis: PropTypes.object,
-	};
+interface Props {
+	duplicatePostId: number;
+	postId: number;
+	postType: string;
+	pressThis: any;
+	siteAdminUrl: string;
+}
+
+type ConnectedProps = ReturnType< typeof mapStateToProps > & typeof mapDispatchToProps;
```

Comment:
Is this the best state-of-the-art way how to type connected props? I wish it was simpler 🙂

Reply:
I have no idea. I chose this way without doing any research based on how I felt was most convenient.

At first I was starting to enumerate the connected properties like this:

```ts
interface ConnctedProps {
	siteId: number;
	…
	trashPost: Function;
	…
}
```

Actually, at first I was mashing it all into `Props` but then realized it would be convenient to split them out. Then I felt like this was all pretty redundant because TypeScript is already inferring the types of `mapStateToProps` and `mapDispatchToProps`. Further, because I was hard-coding in those things it seemed fragile, plus I had issues inferring the types of the dispatch functions and I didn't want to widen them.

Using `ReturnType< typeof mapStateToProps >` and `typeof mapDispatchToProps` gave me the information I wanted (a contract for what `Props` are required from the outside as well as a contract for all props) while retaining the flexibility and stronger inference than typing them out myself.

Of note here is that I'm not intending to impose a contract on `mapStateToProps` and `mapDispatchToProps` but rather to make the existing expectations clear. In my IDE now I can hover over the `CalypsoifyIframe` import or instance and it will show me the list of props it expects.

In what ways would you be looking for it to be simpler? I think that like programming in general, there's lots of stylistic choices that we have available in typing systems and we can often type our code in different ways for a similar outcome.

Reply:
> In what ways would you be looking for it to be simpler?

To me, it would be with a more verbose and easier code to read. So, having separate `StateFromProps` and `DispatchFromProps` could help with that.

However, I see how this can feel redundant and fragile.

I don't have a strong opinion about this, but in this case I'm bit more leaning towards readability rather than optimization.

Reply:
> In what ways would you be looking for it to be simpler?

I don't know. The context is that I'm a TypeScript beginner and this single line introduces a lot of new concepts: `typeof` of a function, the `ReturnType` thing (BTW why is `ReturnType` used on one function and not on the other?), composing two types with `&` operator...

Reply:
> why is ReturnType used on one function and not on the other?

😄 

if you give it a moment's pause you might see why…`mapDispatchToProps` isn't a function. the fun part here is that if you try to use `ReturnType< mapDispatchToProps >` it will warn you of that.

> readability rather than optimization

while the point wasn't optimization the way I'd put it, the purpose was to express clearly what needs to be there.

I'm honestly curious @mmtr about your response. it's funny because in my head I don't care to know on that line what the types are - my concern is about communicating out about the React component in files where it's used. to that end I'm comfortable knowing that "the type of the connected props is the return type of the `mapStateToProps` function combined with the type of the `mapDispatchToProps` object.

what information in this line is confusing that you wish you knew? that is, what are you wanting from this line that it doesn't give you? are you wanting to see all the types inside the connected props? if you compare that to the way we have been doing it where those types simply aren't there, what is rougher about having that information available?

also, are you trying to infer everything from the text in the PR or did you checkout the branch and have the IDE show you the information? I don't think this stuff is nearly as useful if we don't take the IDEs into account as the purpose of these types is automation and communication.

> this single line introduces a lot of new concepts

I can explain what they mean, but I think your point is that there's new stuff. at least Microsoft has [good documentation](https://www.typescriptlang.org/docs/handbook/basic-types.html) for it as well as some intro tutorials. in the end I think we'll most commonly see a smaller subset of syntax in practice. we'll have to learn some new stuff _if we want to add typing_ but I want to stress in all this that we shouldn't feel obliged to introduce types if we aren't excited about it (on an individual developer/module/function/day basis).

 - `ReturnType< T >` means "use the type that the function `T` returns"
 - `typeof T` means "use the same type as T"
 - `T & U` is the type intersection. it means "use a type that combines the elements of the type `T` and also of the type `U`

TypeScript as a _structural_ type system stands in contrast to _nominative_ type systems meaning that it's permissive by nature - as long as a type meets the minimum requirements it will be fine. This isn't like Java where we have to also be of the same type name to work. This code shows some difference…

```ts
interface HasSpeed {
  speed: number;
}

const a: HasSpeed = { speed: 5.2 };
const b = { speed: 13.2, direction: 147 };

function distance( o: HasSpeed, ms: number ) {
  return o.speed * ms * 1000;
}

distance( a, 4.3 ); // obviously fine
distance( b, 3.1 ); // also fine, even though `b` isn't a `HasSpeed` it still meets the contract
```

One implication here is that we often see partial types or constructs that say "hey computer, _you_ figure out the types."

Again, to what purpose are we typing? I want to increase the safety of the code by having the computer automate bug discovery before shipping and I want to communicate to other devs what the expectations are for using my code. React components are of type `Component< P, S >` where we supply the `P` for prop type and `S` for internal state type. This line of code is saying, "TypeScript, join the inferred types of the connected functions and call that the connected props type"

---

Please let me know if my lengthy responses are introducing more confusion than helping.

Reply:
there are a few more generic "type wrappers" that come out of the box in TypeScript

 - `Readonly<T>` specifies that we can't change properties on the type `T`. This, for example, could be used to mark our state tree as read-only and prevent people from accidentally modifying `state` directly (I just saw someone do this maybe even last week in a PR where the write-path was obscured through a couple of variable assignments)

 - `Partial<T>` takes an object type and returns a new type where each property may be missing. it's a shorthand since we can do this by hand.

```ts
interface Vector { x: number; y: number, z: number }
interface PartialVector { x?: number, y?: number, z?: number }
interface PartialVector2 = Partial<Vector>;
```

like the example with `ReturnType<T>` we can see here that these mapping types allow us to keep our actual annotation footprint small and versatile. if we add a `t` element to the `Vector` interface then `PartialVector2` automatically gets it as well but `PartialVector` will be missing it. it's just like normal code except the values are types.

 - `Nullable<T>` just like `Partial<T>` except the properties have to be there, though they can be `null`

 - `Pick<T, K extends key T>`  this one allows us to constrain only a subset of type `T`

```ts
interface Site {
    siteId: number;
    title: string;
    url: string;
}

type SiteLike = Pick < Site, 'siteId' | 'title' >
```

![Screen Shot 2019-04-02 at 4 09 23 PM](https://user-images.githubusercontent.com/5431237/55442017-c1048b80-5561-11e9-9181-c9c4d208fac9.png)

 - `Parameters< T >` returns the type of function arguments for a function of type `T`

there are more of course. this might make a good learnup

Reply:
> what information in this line is confusing that you wish you knew?

just like @jsnajdr said, I'm a TypeScript beginner and this line is introducing some concepts I'm not able to understand by just reading the code (unlike the rest of the PR).

Also, the fact that `mapStateToProps` and `mapDispatchToProps` are defined at the end of the file and not before this line is also contributing to my confusion.

> are you trying to infer everything from the text in the PR or did you checkout the branch and have the IDE show you the information? I don't think this stuff is nearly as useful if we don't take the IDEs into account as the purpose of these types is automation and communication.

yeah, that's true. I didn't checkout the branch and observed the changes on the IDE which definitely are useful there. But that's how I usually review code (I only checkout the branch to test locally, and then read the changes over the PR). 

> Please let me know if my lengthy responses are introducing more confusion than helping.

They are are useful for me. I love having these discussions 🙂. And let me remind what a wonderful job you did here!

Reply:
Today I realized what was behind the gut feeling I had when seeing the `ConnectedProps`. It's the fact that the declaration doesn't contain any new information beyond what is already available in the source:
- the connected component accepts `Props` as input
- the `connect` function has a well-known type: it takes the `Props` passed from outside, adds the output of `mapStateToProps` and of `mapDispatchToProps`, and passes the resulting union down to the wrapped component.

Therefore, the fact that
```
ConnectedProps === Props & ReturnType<mapStateToProps> & ReturnType<mapDispatchToProps>
```
is an obvious universal truth about `connect`. Then why I need to repeat it in every file? With types for `connect` available, it should be inferred, shouldn't it?

Reply:
> Please let me know if my lengthy responses are introducing more confusion than helping.

Your responses are very helpful, thank you!

Reply:
> It's the fact that the declaration doesn't contain any new information beyond what is already available in the source…an obvious universal truth

I've had to ponder this question and how to respond. There are a few ways to look at this. One way is that the type system is forcing us to do something slightly awkward or redundant. Another way to look at it is that we're matching the types with the way we code our components - composed pieces.

If we consider types as _sets_ this line should make sense as we walk through what's happening. But first, let's point out the _elephant in the room_: we can skip all this.

```ts
interface Props {
  siteId: number;
  postId: number;
}

export default connect(
	( state, { siteId, postId }: Props ) => ( { … } ),
	{ setPostTitle },
)( class extends React.Component {
	…
} )
```

With this formulation TypeScript should infer everything and we wouldn't need the redundancy. We could be writing our code like this but we separate out connected and unconnected components for things like testability. The implications of these separations bleed into the type system too as we have to or want to maintain the expectations on the unwrapped versions of the components.

The last commit where I added the `ProtectFormProps` can help illustrate this.

What is the _type_ of `CalypsoIframe`? _Because_ we use props from the `mapStateToProps` then we know that the type has to include those things beyonds what are defined in the public `Props` (or possibly we could call it `ExternalProps`) interface. The innermost component needs to constrain the input parameters to all things that it uses internally: the full set union of external props, connected props, and protected-form props.

I've created `ConnectedProps` as a convenience to group `connect()` elements but we could inline it just the same.

The `CalypsoIframe` component then is the set union of these types (for _reasons_ we use the intersection operator) or…
```ts
Component<ExternalProps & ConnectedProps & ProtectFormProps, InternalState>
```

Notice that we're _not_ adding any type annotations to `connect()`? Why? because we started on the inside and typed the inside and let TypeScript do the unwrapping. `protectForm` performs a kind of set difference on the types… (please excuse my made-up type syntax below)
```haskell
protectForm :: Component (P ∪ ProtectFormProps) -> Component P
```

That is, we're lifting out the subset of the type that is also `ProtectFormProps`. It leaves exposed `ExternalProps & ConnectedProps`. `connect()` does the same thing by removing the `ConnectedProps` though at the point it gets to `connect()` it's not using this convenience function but rather inferring them on its own.

We're not starting at `connect()` and then somehow injecting props into `CalypsoIframe`. Rather, we're starting with `CalypsoIframe` and the props it needs then using `connect()` to fulfill _some_ of them, leaving the rest to be supplied from the outside.

---

Let's backup and repeat in a different way:

 - `CalypsoIframe` actually needs things like `markChanged()` as provided by `protectForm()`. That means it _has_ to include that in its type.
 - Wrappers and higher-order components might _supply_ some props while _removing_ others. They do this through set operations on the types.

This is good news because when we do things like type `protectForm` the way I have here we get a clear type telling us what we should pass in on the imported object regardless of the internal implementation details, though we can still unwrap the internals for testing and non-standard composition.

We probably don't want `protectForm()` operating in our tests, but `this.props.markChanged()` is still called from inside the component. The inner type reminds us to provide mock functions for it in those cases.

![WrappedTypes](https://user-images.githubusercontent.com/5431237/55595654-c5fc4300-56f9-11e9-8805-3278698e898c.gif)

In this screencast we can see how it complains if we try and set `markChanged()` manually. While working this PR I discovered that if we _do_ pass in a value for `markChanged()` it will end up replacing the one built-in to the `protectForm()` wrapper. Is this a bug in the wrapper? I don't know, but now I am aware of that behavior and can enforce through the types not to do it.

If we pass in `siteId` to the connected component it will likewise complain, which is good, because that passed-in `siteId` will get overwritten by the one in `mapStateToProps`. This would be a latent bug in production otherwise and only expose itself when we have something happen like a mismatch in local and API data - maybe during the time an API call is resolving.